security/fma-l1-upgrades.md

```diff
@@ -0,0 +1,306 @@
+# [Project Name]: Failure Modes and Recovery Path Analysis
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+
+- [Introduction](#introduction)
+- [Failure Modes and Recovery Paths](#failure-modes-and-recovery-paths)
+  - [[Name of Failure Mode 1]](#name-of-failure-mode-1)
+  - [[Name of Failure Mode 2]](#name-of-failure-mode-2)
+- [Audit Requirements](#audit-requirements)
+- [Action Items](#action-items)
+- [Appendix](#appendix)
+  - [Appendix A: This is a Placeholder Title](#appendix-a-this-is-a-placeholder-title)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+_Italics are used to indicate things that need to be replaced._
+
+|                    |            |
+| ------------------ | ---------- |
+| Author             | Maurelian  |
+| Created at         | 2024-03-26 |
+| Initial Reviewers  | [TBD]      |
+| Need Approval From | [TBD]      |
+| Status             | Draft      |
+
+## Introduction
+
+## Introduction
+
+This document covers the L1 Upgrades project, which aims to improve and standardize the approach to upgrading L1 contracts across the superchain. The project introduces a new upgrade system that consolidates ProxyAdmin contracts, implements a two-step upgrade flow, and provides a more deterministic and safer upgrade process.
+
+Below are references for this project:
+
+- [L1 Upgrades Design Doc](../protocol/l1-upgrades.md)
+- [OPCM spec](https://github.com/ethereum-optimism/specs/blob/opcm/upgrades/specs/experimental/op-contracts-manager.md)
+
+### FM1: Contract Initializers
+
+#### **Description:**
+
+During the first step of the ugprade, the contracts will not yet have a shared
+storage location for the `initialized` value. Therefore each contract will require custom storage
+manipulation to properly reset the `initialized` value, without modifying values which may be
+packed into the same slot.
```

Comment:
I was looking into this more and realized a few more things. Here's my summary below:

1. This will be a one time operation whereby we will move all contracts that are initializable to the new OpenZepplin Initializable [v5.1.0 contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.1.0/contracts/proxy/utils/Initializable.sol).
2. Some contracts use older versions of [Initializable.sol](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/release-v4.4/contracts/proxy/utils/Initializable.sol) which do not use unstructured storage. We want all contracts to use the unstructured storage because it's better to have a uniform way to reset initializer values across the entire system. 
3. When moving to the new  [v5.1.0 contract](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.1.0/contracts/proxy/utils/Initializable.sol), we need to be mindful about storage slot collisions. It'll be necessary to use spacers in the implementation contracts to ensure storage collisions are avoided.

Reply:
> For my own understanding, what's an example of a contract that we use that isn't using unstructured storage for the initialized value?

All of them. :)

ie. they are using [this](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/v4.7.3/contracts/proxy/utils/Initializable.sol#L62) version of `Initializable`, which just puts `_initialized` in the [first slot](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/v4.7.3/contracts/proxy/utils/Initializable.sol#L62).

The proposal is to use [this version](https://github.com/OpenZeppelin/openzeppelin-contracts-upgradeable/blob/fa525310e45f91eb20a6d3baa2644be8e0adba31/contracts/proxy/utils/Initializable.sol#L77).

Reply:
Awesome, thanks for the clarification. Going to resolve this.

Reply:
oh, I didn't see your initial reply to self when I sent mine. 

> When moving to the new v5.1.0 contract, we need to be mindful about storage slot collisions. It'll be necessary to use spacers in the implementation contracts to ensure storage collisions are avoided.

Yes, FWIW I have recently dealt with this pretty extensively in the [Isthmus PR](https://github.com/ethereum-optimism/optimism/pull/12932/files#diff-7d99abb08e3cb939c50026beb0d42dc3b72c3e3b5c29e254ebd1e63c552e25bf) that was not included.